flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/FileUploads.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.handler;
+
+import org.apache.flink.util.FileUtils;
+import org.apache.flink.util.Preconditions;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * A container for uploaded files.
+ *
+ * <p>Implementation note: The constructor also accepts directories to ensure that the upload directories are cleaned up.
+ * For convenience during testing it also accepts files directly.
+ */
+public final class FileUploads implements AutoCloseable {
+	private final Collection<Path> directoriesToClean;
+	private final Collection<Path> uploadedFiles;
+
+	@SuppressWarnings("resource")
+	public static final FileUploads EMPTY = new FileUploads();
+
+	public static FileUploads forDirectory(Path directory) throws IOException {
+		final Collection<Path> files = new ArrayList<>(4);
+		Preconditions.checkArgument(directory.isAbsolute(), "Path must be absolute.");
+		Preconditions.checkArgument(Files.isDirectory(directory), "Path must be a directory.");
+
+		FileAdderVisitor visitor = new FileAdderVisitor();
+		Files.walkFileTree(directory, visitor);
+		files.addAll(visitor.getContainedFiles());
+		
+		return new FileUploads(Collections.singleton(directory), files);
+	}
+
+	private FileUploads() {
+		this.directoriesToClean = Collections.emptyList();
+		this.uploadedFiles = Collections.emptyList();
+	}
+
+	public FileUploads(Collection<Path> directoriesToClean, Collection<Path> uploadedFiles) {
+		this.directoriesToClean = Preconditions.checkNotNull(directoriesToClean);
+		this.uploadedFiles = Preconditions.checkNotNull(uploadedFiles);
```

Comment:
I know it's really nitpicking what I'm doing here, but I think it would be slightly better to let FileUploads only represent the upload directories and add a method FileUploads#getFiles which returns a Collection<File> which are all files being found in the upload directory. The difference is that we don't initialize FileUploads with it. That would effectively enforce that all files reside in the given upload directories. What we could do now is to initialize this class with directories /web/upload/a, /web/upload/b and files /web/different/path/file where the files are somewhere else located. Due to this, we not only need to delete the directories but also all files.